### ISSUE TITLE
Too many hydration / lineage functions

### ISSUE THREAD

**Issue by estellecomment** [medic/medic-webapp#3826]
Api, sentinel and webapp each have their separate hydration functions. That's dangerous duplication (it's kinda tricky code, it's already had bugs).

Merge it all together, either in an npm module or if we merge repos, in a node module accessible by path.

**Comment by SCdF** [comment]
We should take account of this at the same time: https://github.com/medic/medic-webapp/issues/3913

**Comment by SCdF** [comment]
(I renamed the title because I'm sick of searching for this ticket with 'lineage' and not finding anything :shipit: )

**Comment by rmhowe** [comment]
**Progess so far (2/3/2018):**
**Done**
- Added lineage shared library (converted to es5 and using promises, with tests converted to mocha with promises)
- Replaced:
  - Uses of lineage library within sentinel
  - The `places` controller within api
  - The `export-data` and `export-data-2` services within api
  - The `lineage-model-generator` service within webapp

**To do**
- Fix the `places` controller tests and `export-data` service tests
- Copy the `getPatientContactUuid` function from `sentinel/lib/utils.js` for use in the lineage library

PR here: https://github.com/medic/medic-webapp/pull/4256

**Comment by SCdF** [comment]
Please review @garethbowen, apologies that the PR is so massive!

**Comment by SCdF** [comment]
(If you're cool with it leave the merging to me)

**Comment by garethbowen** [comment]
Back to you. Don't merge until the `2.15.x` is taken.

**Comment by SCdF** [comment]
Since 2.15 is nearly completion I'm going to wait (instead of merging and creating a 2.15 branch. Hopefully that turns out to be the right decision!

**Comment by SCdF** [comment]
AT: all the things. In all seriousness this effects all places where lineages are dealt with, which covers webapp, api and sentinel.

Mapping code changes roughly to features we get
```
./api/controllers/people.js:
 - /api/v1/people
 - extract-person-contacts migration
 - places and users (see below)
./api/controllers/places.js:
 - /api/v1/places
./api/controllers/users.js:
 - /api/v1/users
./api/services/export-data-2.js:
 - /api/v2/export
./api/services/export-data.js:
 - /api/v1/export
./sentinel/schedule/due_tasks.js:
 - The due tasks schedule
./sentinel/transitions/index.js:
 - All transitions, but specifically ones here:
./sentinel/transitions/multi_report_alerts.js:
./sentinel/transitions/registration.js:
./sentinel/transitions/update_clinics.js:
./static/js/services/lineage-model-generator.js:
 - Creating and editing contacts
 - Adding reports
 - Displaying messages
 - General enketo rendering
 - Select2 searches
```

**Comment by ngaruko** [comment]
The `export` feature now breaks the app, with this error (mainly `api/v1/export/contacts`, `api/v1/export/messages`, `api/v1/export/feedback`, `api/v1/export/logs`, `api/v1/export/audit`):
```1 - admin [16/Mar/2018:02:09:24 +0000] "GET /api/v1/export/messages?format=xml&locale=en HTTP/1.1" - - "http://localhost:5988/medic/_design/medic/_rewrite/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/65.0.3325.146 Safari/537.36"
UNCAUGHT EXCEPTION!
Server error:
 { TypeError: Cannot read property 'then' of undefined
    at /Users/bede/.horticulturalist/deployments/medic-api/md5-nAGDADc5Elc5Eo4ok5REFg==/controllers/export-data.js:65:9
    at executeBound (/Users/bede/.horticulturalist/deployments/medic-api/md5-nAGDADc5Elc5Eo4ok5REFg==/node_modules/underscore/underscore.js:701:67)
    at bound (/Users/bede/.horticulturalist/deployments/medic-api/md5-nAGDADc5Elc5Eo4ok5REFg==/node_modules/underscore/underscore.js:733:14)
    at /Users/bede/.horticulturalist/deployments/medic-api/md5-nAGDADc5Elc5Eo4ok5REFg==/auth.js:131:18
    at executeBound (/Users/bede/.horticulturalist/deployments/medic-api/md5-nAGDADc5Elc5Eo4ok5REFg==/node_modules/underscore/underscore.js:701:67)
    at bound (/Users/bede/.horticulturalist/deployments/medic-api/md5-nAGDADc5Elc5Eo4ok5REFg==/node_modules/underscore/underscore.js:733:14)
    at /Users/bede/.horticulturalist/deployments/medic-api/md5-nAGDADc5Elc5Eo4ok5REFg==/auth.js:95:18
    at Request._callback (/Users/bede/.horticulturalist/deployments/medic-api/md5-nAGDADc5Elc5Eo4ok5REFg==/auth.js:19:5)
    at Request.self.callback (/Users/bede/.horticulturalist/deployments/medic-api/md5-nAGDADc5Elc5Eo4ok5REFg==/node_modules/request/request.js:186:22)
    at emitTwo (events.js:126:13)
  domain: 
   Domain {
     domain: null,
     _events: { error: [Function] },
     _eventsCount: 1,
     _maxListeners: undefined,
     members: [] },
  domainThrown: true }```

In some cases, it also prompt for a password. If this is not caused by this issue, I will create a new issue but for the time being I will push this back and let the devs decide. Aye @garethbowen @SCdF ?

**Comment by SCdF** [comment]
Whoops, sorry about that, shoddy merging on my part!

I've fixed this and am going to do a deeper round of testing, as I've found a couple of other interesting bits and bobs. If you can keep testing the rest of it that would be greatly appreciated :-)

**Comment by SCdF** [comment]
> this bring it down with a 502 Bad Gateway response

One of the issues that I've fixed (not yet merged) would cause API to restart. This was due to applying a V2 optimisation to the V1 code path. I've changed that now, so there shouldn't be any way to kill api now (that we know of anyway)

**Comment by ngaruko** [comment]
`Choice` @SCdF

**Comment by SCdF** [comment]
@garethbowen please review https://github.com/medic/medic-webapp/pull/4307

You'll note there are no tests for the lineage changes. I don't think these tests are scaling or even necessarily all testing what they think they are testing. I wrote #4306 to deal with this, and once kanso is gone I'm going to write a suite of integration tests that tests this critical (and historically buggy AF) piece of our codebase against a real DB.

**Comment by ngaruko** [comment]
LGTM. ```export server logs` is still broken and errors but there is an issue for that [here](https://github.com/medic/medic-webapp/issues/4211)
